### ISSUE TITLE
Pureconfig errors are misleading in the face of ConfigConverts missing.

### ISSUE THREAD

**Issue by coltfred** [pureconfig/pureconfig#35]
Today if you have nested configs like the following

```
case class Foo(bar: Bar)
case class Bar(nel: NonEmptyList[String])
```

When you try and load a config for `Foo` you get an error like this:
`could not find implicit value for parameter conv: pureconfig.ConfigConvert[Foo]` when in reality, it's Bar that's the problem.

Do you have any ideas about how we could improve the error? One thought I had was to add a "semi automatic" derivation mode where we could create a `ConfigConvert` on the companion object of each of the nested objects.

Like this:

```
case class Bar(nel: NonEmptyList[String])

object Bar {
  implicit val configConvert = pureconfig.derive[Bar]
}

case class Foo(bar:Bar)

object Foo{
  implicit val configConvert = pureconfig.derive[Foo]
}
```

Obviously this changes the way that pureconfig works and we'd have to have 2 different imports, one for "full auto" and one for "semi auto" derivation.

@melrief, Would you be interested in a change like this? Do you see another way around the terrible error messages?

Thanks!

**Comment by melrief** [comment]
The problem with errors on typeclasses that can be derived from other typeclasses is that Scala gives you an error only for the top level type because it doesn't understand that the reason why it cannot use an instance is that one or multiple fields of a type don't have an instance themselves causing the error. I once asked this same [question](http://stackoverflow.com/questions/32877019/make-scalas-implicitnotfound-annotation-more-precise) about errors and I discovered that there is no easy way to get a good error message. To be precise, I made a small prototype using very very black magic code using Shapeless and I've been able to improve the precision of error messages but the code needs a lot of work before it will be usable and can make things even worse.

The style that you propose, to have `auto` and `semiauto`, is similar to one of my favourite libraries named [circe](https://github.com/travisbrown/circe) and I think it's a great idea. If you want to work on a PR I would be more than happy of commenting and merging it.

**Comment by coltfred** [comment]
@melrief Thanks for taking the time to respond. My mental model was also based on Circe so I'm glad we're on the same page. I'll try and take a stab at it, though it'll probably be a week or so before I can carve out the time.

**Comment by francisdb** [comment]
I just tested how the play-json format macro handled this and they seem to be able to provide more targeted messages. They however ask you to work in 2 steps by first generating the converter and then converting the json ast

the config case class looks like this

``` scala
case class BaseConfiguration(
  nodeServerAddress: String,
  luceneDirectoryIndex: Option[Path],
  defaultTimeZone: TimeZone
)
```

``` scala
val reads = Json.format[BaseConfiguration]
// above line fails with:
// No implicit format for Option[java.nio.file.Path], java.util.TimeZone available
val result = json.validate[BaseConfiguration]
```

where for pureconfig I get

``` scala
val result: Try[BaseConfiguration] = loadConfig[BaseConfiguration]
// could not find implicit value for parameter conv: pureconfig.ConfigConvert[foo.BaseConfiguration]
```

**Comment by melrief** [comment]
I'm not sure how play-json is creating the error message and searching for the string "No implicit format for" in their [github repo](https://github.com/playframework/playframework/search?utf8=%E2%9C%93&q=%22No+implicit+format+for%22&type=Code) doesn't return any result.
For pureconfig, we create a case class converter only if there are all the converters for the fields of the class. The main effect of this is that when the implicit derivation fails, the system just returns error without knowing what field made it fail because it doesn't know which instance is the "right" one in the first place and what are the requirements that are missing to make it work. There are some techniques that I'm exploring to get better error but, without creating a macro specialized for pureconfig, they can get very complex.

**Comment by francisdb** [comment]
The macro code generating the error can be found here:
https://github.com/playframework/playframework/blob/2.5.x/framework/src/play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala#L196

It seems that for play > 2.5 this message will be even more detailed
https://github.com/playframework/playframework/blob/master/framework/src/play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala#L417

**Comment by melrief** [comment]
@francisdb thanks for the links, I'll have a look at the code. I think we could change the codec generator from the one that we have shapeless based to a custom macro based one. I'm not happy to have my own macro but if it makes much cleaner the error messages then I'm for trying it. I'll talk with the other two authors and see how we want to proceed.

**Comment by melrief** [comment]
#279 should fix this

**Comment by ruippeixotog** [comment]
Closed by #279.
